### PR TITLE
`json_typeof` and `jsonb_typeof` function

### PR DESCRIPTION
**Pull request by anna-ahmed19**
Added the `json_typeof` and `jsonb_typeof` functions under [Issue#4216](https://github.com/diesel-rs/diesel/issues/4216)

### CODE REVIEW

File: diesel/src/pg/expression/functions.rs [hunk 1]
```diff
@@ -1634,3 +1636,113 @@ define_sql_function! {
...
+    /// let result = diesel::select(json_typeof::<Json, _>(json!(true)))
+    ///     .get_result::<String>(connection)?;
+    ///
+    /// assert_eq!("boolean".to_string(), result);
+    ///
+    /// let result = diesel::select(json_typeof::<Json, _>(json!(null)))
+    ///     .get_result::<String>(connection)?;
+    ///
+    /// assert_eq!("null".to_string(), result);
+    /// #     Ok(())
```
Comment (guissalustiano) [inline diesel/src/pg/expression/functions.rs:1689]: Could you add a test with None input?
It is something like this:
```
    /// let result = diesel::select(json_typeof::<Nullable<Json>, _>(None::<Value>))
    ///     .get_result::<String>(connection)?;
    ///
    /// assert_eq!(<something>, result);
```

File: diesel/src/pg/expression/functions.rs [hunk 2]
```diff
@@ -1634,3 +1636,113 @@ define_sql_function! {
...
+    ///
+    /// let result = diesel::select(jsonb_typeof::<Jsonb, _>(json!(true)))
+    ///     .get_result::<String>(connection)?;
+    ///
+    /// assert_eq!("boolean".to_string(), result);
+    ///
+    /// let result = diesel::select(jsonb_typeof::<Jsonb, _>(json!(null)))
+    ///     .get_result::<String>(connection)?;
+    ///
+    /// assert_eq!("null".to_string(), result);
```
Comment (guissalustiano) [inline diesel/src/pg/expression/functions.rs:1743]: Could you also add a None test here?

**Review by guissalustiano — CHANGES_REQUESTED** [review]
Thanks for opening the PR!

The changes look good, I only would like to see some tests about the null behavior, in past cases, Postgres returns `NULL` when you pass `NULL` to this kind of functions.
If this is case, the function signature is wrong, It should also cover the NULL input return.

To fix that we can use 
```rust
fn json_typeof<E: JsonOrNullableJson + SingleValue + MaybeNullableValue<Text>>(e: E) -> E::Out;
```
The `E: MaybeNulllableValue<Text>` says that the output values has the same nullability as the input value. That means Nullable<Json<T>> evaluates to Nullable<Text> and Json<T> evaluates to Text via the Arr::Out return type

**Comment by anna-ahmed19** [comment]
Thank you @guissalustiano for the suggestions. 
I have added `MaybeNullableValue<Text>` and the `None` tests. 
Let me know any further changes.

**Review by guissalustiano — APPROVED** [review]
Thanks!

**Comment by weiznich** [comment]
Thanks for submitting this PR :heart:
